wallet-new/src/Cardano/Wallet/API/V1/Migration/Types.hs

```diff
@@ -121,6 +122,11 @@ instance Migrate V0.CCoin (V1 Core.Coin) where
 instance Migrate (V1 Core.Coin) V0.CCoin where
     eitherMigrate (V1 c) = pure (V0.encodeCType c)
 
+instance (n ~ m, ValidMnemonicSentence n)
+    => Migrate (Mnemonic n) (V0.CBackupPhrase m) where
```

Comment:
Out of curiosity, any reason to distinguish `n` from `m` here ?

Reply:
I was wondering this as well (on another hand I rarely use this machinery so don't know how it exactly works)

Reply:
It's a trick to get better type inference. The `n ~ m` constraint is only checked after the instance is selected. Suppose we have some polymorphic code like:

```
foo :: Mnemonic n -> Int
bar :: CBackupPhrase 9

mnemonic <- convert bar
pure (foo mnemonic)
```

Because `n` is polymorphic here, we can't select the instance `Migrate (Mnemonic n) (CBackupPhrase n)`, because we don't know that these are the same type. So we say "No instance found for `Migrate (Mnemonic n0) (CBackupPhrase 9)`.

When we use the`n ~ m` constraint and two type variables, this allows GHC to select the instance head `Migrate (Mnemonic n) (CBackupPhrase m)`, and after it has selected this instance, it will check the context: `(n ~ m, ValidMnemonicSentence n)`. The `n ~ m` constraint tells GHC that `n` must be equal to `m`, which means that the more specialized type of the instance is `Migrate (Mnemonic m) (CBackupPhrase 9)` and becomes `Migrate (Mnemonic 9) (CBackupPhrase 9)`.

This is a good blog post on the matter: https://chrisdone.com/posts/haskell-constraint-trick